src/basic/exec-util.c

```diff
@@ -71,24 +71,30 @@ static int do_execute(
                 gather_stdout_callback_t const callbacks[_STDOUT_CONSUME_MAX],
                 void* const callback_args[_STDOUT_CONSUME_MAX],
                 int output_fd,
-                char *argv[]) {
+                char *argv[],
+                ExecDirFlags flags) {
 
         _cleanup_hashmap_free_free_ Hashmap *pids = NULL;
         _cleanup_strv_free_ char **paths = NULL;
         char **path;
         int r;
+        bool parallel_execution;
 
         /* We fork this all off from a child process so that we can somewhat cleanly make
          * use of SIGALRM to set a time limit.
          *
-         * If callbacks is nonnull, execution is serial. Otherwise, we default to parallel.
+         * We attempt to perform parallel execution if configured by the user, however
+         * if `callbacks` is nonnull, execution must be serial.
```

Comment:
I think an `assert()` in this case would be better than silently changing the required behavior, this is an interface violation after all.

Reply:
I think failing hard with an assertion would not be the right behavior. Yes, combining callbacks with the PARALLEL flag won't work, but it also will not result in any kind of undefined state withing the program - it will just execute the tasks slightly slower.
That's why I would read the PARALLEL flag as "make it parallel if possible" and not as "the scripts *must* be run in parallel".
If I would use an assertion here, the penalty of making a slightly wrong assumption about the API feels overly harsh (crashing the whole application, maybe even in a really bad place (this code is used very early for generators etc. as well)).